src/data/nsfg.clj

```diff
@@ -1,3 +1,165 @@
+
+(defn reader
+  "Open path with io/reader; coerce to a GZIPInputStream if suffix is .gz"
+  [path]
+  (if (.endsWith path ".gz")
+    (io/reader (GZIPInputStream. (io/input-stream path)))
+    (io/reader path)))
+
+(defn read-dictionary-data
+  "Parse lines from `rdr` according to the specification in `dict`.
+   Return a lazy sequence of parsed rows."
+  [dict rdr]
+  (let [parse-fn (make-row-parser dict)]
+    (map parse-fn (line-seq rdr))))
+
+(defn in?
+  "true if coll contains elm"
+  [coll elm]
+  (some #(= elm %) coll))
+
+(defn clean-fem-preg
+  " Recodes variables from the pregnancy frame.
+
+    # mother's age is encoded in centiyears; convert to years
+    df.agepreg /= 100.0
+
+    # birthwgt_lb contains at least one bogus value (51 lbs)
+    # replace with NaN
+    df.loc[df.birthwgt_lb1 > 20, 'birthwgt_lb1'] = np.nan
+
+    # replace 'not ascertained', 'refused', 'don't know' with NaN
+    na_vals = [97, 98, 99]
+    df.birthwgt_lb1.replace(na_vals, np.nan, inplace=True)
+    df.birthwgt_oz1.replace(na_vals, np.nan, inplace=True)
+
+    # birthweight is stored in two columns, lbs and oz.
+    # convert to a single column in lb
+    # NOTE: creating a new column requires dictionary syntax,
+    # not attribute assignment (like df.totalwgt_lb)
+    df['totalwgt_lb'] = df.birthwgt_lb1 + df.birthwgt_oz1 / 16.0"
+  [ds]
+  (-> ds
+      (tc/map-rows (fn [{:keys [agepreg]}]
+                     {:agepreg (when-not (nil? agepreg)
+                                 (float (/ agepreg 100)))}))
+      (tc/map-rows (fn [{:keys [birthwgt-lb]}]
+                     (let [na-vals [51 97 98 99]]
+                       {:birthwgt-lb (if (in? na-vals birthwgt-lb)
+                                       nil
+                                       birthwgt-lb)})))
+      (tc/map-rows (fn [{:keys [birthwgt-oz]}]
+                     (let [na-vals [97 98 99]]
+                       {:birthwgt-oz (if (in? na-vals birthwgt-oz)
+                                       nil
+                                       birthwgt-oz)})))
+      (tc/map-rows (fn [{:keys [birthwgt-lb birthwgt-oz]}]
+                     {:totalwgt-lb (when (and birthwgt-lb birthwgt-oz)
+                                     (-> (/ birthwgt-oz 16)
+                                         float
+                                         (+ birthwgt-lb)))}))))
+
+(defn read-fem-preg-dataset
+  "Read Stata data set, return an dataset."
+  ([]
+   (read-fem-preg-dataset "resources/2002FemPreg.dct" "resources/2002FemPreg.dat.gz"))
+  ([dict-path data-path]
+   (let [dict   (read-dictionary dict-path)
+         header (map (comp keyword :name) dict)]
+     (with-open [r (reader data-path)]
+       (->> (tc/dataset (read-dictionary-data dict r)
+                        {:layout :as-rows
+                         :column-names header
+                         :dataset-name "2002FemPreg"})
+            (clean-fem-preg))))))
+
+(defn get-column-frequency-by-index [ds col index]
```

Comment:
Nice!

It would be more efficient to simply filter by the index and count.

But I see that you wish to make things more similar to the Python code, so maybe it makes sense to use Frequencies as you did.